SOHO-7580 - Merge Searchfield and ToolbarSearchfield APIs

EdwardCoyle opened the pull request:
> Explain the **details** for making this change. What existing problem does the pull request solve?

There were previously two separate APIs for a Searchfield that was implemented in the context of a Toolbar.  The implementation was messy and caused many issues when it came to resizing the Searchfield, listening to events, and sharing common logic between the two.  This PR is an attempt at cleaning up both APIs with goals of:

- reducing total amount of code between both APIs
- improving the common API to be more easily tested
- allowing certain features on regular Searchfield that were previously limited to Toolbar Searchfield (fx: expand/collapse)
- Prepare the Searchfield for shared usage between the existing Toolbar component and the new Flex Toolbar component
- introducing functional/e2e testing on this component (previously had none)

> **Related issue (required)**:

https://jira.infor.com/browse/SOHO-7580

> **Items left to do**

- [x] Merge JS APIs for both components into standalone Searchfield
- [x] Cleanup lifecycle issues that crop up as a result of the API merges
- [x] Smooth out events
- [x] Merge the CSS libraries for both components, and remove extraneous CSS
- [x] Add functional tests
- [x] Add e2e tests

> **Steps necessary to review your pull request (required)**:

Pull this branch and run the demoapp:
- Ensure that all normal Searchfield visual tests work without console errors or visual artifacts: http://localhost:4000/components/searchfield/list
- Ensure that all Toolbar Searchfield visual tests work without console errors or visual artifacts: http://localhost:4000/components/toolbarsearchfield/list
- Check that non-collapsible searchfields in the Flex Toolbar visual tests are working as expected:http://localhost:4000/components/toolbar-flex/list

Also, run the functional and e2e tests for:
- autocomplete
- popupmenu
- searchfield (new)

> Other Details:

Closes SOHO-7580
Closes SOHO-4820

<!-- After submitting your PR, please check back to make sure tests pass on Travis. -->

tmcconechy commented:
I think this PR needs a merge from master. Seems to have some of my changes in it from earlier...

tmcconechy commented on TODO.md:
```diff
@@ -0,0 +1,18 @@
+# TODO List (SOHO-7580)
```
Lets remove this before merge / use code comments maybe

EdwardCoyle commented:
@tmcconechy I think I've addressed all the points you brought up except the one about Flex Toolbar.. I didn't add anything new around Flex Toolbar because it's out of scope, and this issue's already big enough.

I've removed the "Do not merge" warning from this PR... I think it's stable enough to get pushed to master.  Would appreciate some review from everyone to make sure nothing crazy's there.

EdwardCoyle commented:
@d-g-h fixed

tmcconechy commented:
See a couple smaller issues.

1) http://localhost:4000/components/searchfield/example-index.html
- type something and select a value
- leave the field
- hit the X
Problem: Every time you hit the x it opens the list

2) http://localhost:4000/components/searchfield/example-categories-and-go-button.html#
is there a way / setting where it wont change size? wondering...

3) http://localhost:4000/components/searchfield/example-categories-short.html
- focus in chrome
- focus state looks a bit funny on the right side

4) http://localhost:4000/components/searchfield/example-context-search-style.html
- not sure if the focus state is correct here. just a box shadow,

5) http://localhost:4000/components/searchfield/test-categories.html
- the button on the right (collapsed with the binoculars) next to button 1
- when you click it it expands and focuses the category. maybe it should focus the text as it takes a second click

6) http://localhost:4000/components/searchfield/test-single-category.html
- same problem on focus state on the right

EdwardCoyle commented:
@tmcconechy regarding:

> 2. http://localhost:4000/components/searchfield/example-categories-and-go-button.html#
> is there a way / setting where it wont change size? wondering...

no, no way to do that right now.  Maybe something we can add later
